### DISCUSSION TITLE
[interface] how to return an interface in data-fetcher?

### DISCUSSION THREAD

**Discussion by jessepinkman9900** [Netflix/dgs-framework#151] [Q&A] [answered]
Hi, 
I'm working with a graphql schema that has interfaces. I'm having trouble returning them in the data-fetcher.
Looking for a worked-out example.
Getting this error in spring logs when I try to execute query in graphiql:
```
graphql.AssertException: You have asked for named object type 'Entity' but its not an object type but rather a 'graphql.schema.GraphQLInterfaceType'
```
I have read: https://netflix.github.io/dgs/advanced/type-resolvers-for-abstract-types/

```graphql
# original schema
type Query {
    movies: [Movie]
}

interface Movie {
    title: String
}

type ScaryMovie implements Movie {
    title: String
    gory: Boolean
    scareFactor: Int
}

type ActionMovie implements Movie {
    title: String
    nrOfExplosions: Int
}
```
```java
@DgsComponent
public class MovieDataFetcher {
    @DgsData(parentType = "Query", field = "movies")
    public List<Movie> movies() {
        return Lists.newArrayList(
                new ActionMovie("Crouching Tiger", 0),
                new ActionMovie("Black hawk down", 10),
                new ScaryMovie("American Horror Story", true, 10),
                new ScaryMovie("Love Death + Robots", false, 4)
            );
    }
}
```
They have solved for a case when the interface has __ONLY ONE__ attrribute, i.e `title` in case for `interface Movie`
```java
@DgsTypeResolver(name = "Movie")
public String resolveMovie(Movie movie) {
    if(movie instanceof ScaryMovie) {
        return "ScaryMovie";
    } else if(movie instanceof ActionMovie) {
        return "ActionMovie";
    } else {
        throw new RuntimeException("Invalid type: " + movie.getClass().getName() + " found in MovieTypeResolver");
    }
}
```

Looking for a solution where Movie had more than one attribute.
For _example_ __IF__ schema was
```graphql
# example schema
type Query {
    movies: [Movie]
}

interface Movie {
    title: String
    year: Int
    director: String
}

type ScaryMovie implements Movie {
    title: String
    gory: Boolean
    scareFactor: Int
}

type ActionMovie implements Movie {
    title: String
    nrOfExplosions: Int
}
```
__THEN__ what would the resulting `@DgsData` or `@DgsTypeResolver` be?

**Comment by srinivasankavitha** [comment, marked as answer]
The DgsTypeResolver would stay the same. The types ActionMovie and ScaryMovie would define the additional fields, i.e, year and director ( title is already there). The datafetcher can create an instance of either ActionMovie or ScaryMovie with the additional fields populated by the same datafetcher.

You can also define field specific data fetchers if you are not able to populate all the fields in the same movies data fetcher. graphql-java will take care of calling every data fetcher for the fields you specify.

Hope that helps.

**Reply by jessepinkman9900** [reply]
This works. Thank you! :)
